src/daemon/daemon.cpp

```diff
@@ -3701,60 +3588,73 @@ void mp::Daemon::populate_instance_info(VirtualMachine& vm,
+    mp::NetworkInterface new_if{br_interface, generate_unused_mac_address(allocated_mac_addrs), true};
+    mpl::log(mpl::Level::debug,
+             category,
+             fmt::format("New interface {{\"{}\", \"{}\", {}}}", new_if.id, new_if.mac_address, new_if.auto_mode));
 
-            for (const auto& command : commands)
-            {
-                mpu::run_in_ssh_session(session, command);
-            }
-        }
-        catch (const mp::SSHExecFailure&) // In case there is an error executing the command, report it.
-        {
-            // Currently, the only use of running commands at boot is to configure networks. For this reason,
-            // the warning shown here refers to that use. In the future, in case of using the feature for other
-            // purposes, it would be necessary to add a description or a failure message to show if the
-            // execution fails.
-            if (!warned_exec_failure)
-            {
-                add_fmt_to(warnings,
-                           "Failure configuring network interfaces in {}: is Netplan installed?\n"
-                           "You can still configure them manually.\n",
-                           name);
-            }
+    // Add the new interface to the spec.
+    spec.extra_interfaces.push_back(new_if);
 
-            warned_exec_failure = true;
-        }
-        catch (const ExitlessSSHProcessException& e) // Timeout does not mean there is a configuration error.
-        {
-            mpl::log(mpl::Level::warning,
-                     category,
-                     fmt::format("Exception when executing command at boot in {}: {}", name, e.what()));
-        }
-        catch (const SSHException&) // The SSH session could not be created.
-        {
-            add_fmt_to(warnings,
-                       "Cannot create a SSH shell to execute commands on {}, you can configure new\n"
-                       "interfaces manually via Netplan once logged in to the instance.\n",
-                       name);
-        }
+    mpl::log(mpl::Level::trace, category, "Prepare networking");
+    config->factory->prepare_networking(spec.extra_interfaces);
+    new_if = spec.extra_interfaces.back(); // prepare_networking can modify the id of the new interface.
+    mpl::log(mpl::Level::trace, category, fmt::format("Done preparation, new interface id is now \"{}\"", new_if.id));
 
-        commands.clear();
-        persist_instances();
+    // Add the new interface to the VM.
+    mpl::log(mpl::Level::trace, category, "Adding new interface to instance");
+    try
+    {
+        instance->add_network_interface(spec.extra_interfaces.size() - 1, new_if);
+        mpl::log(mpl::Level::trace, category, "Done adding");
+
+        // Configure the new interface via cloud-init (this won't throw).
+        std::vector<mp::NetworkInterface> interfaces_to_add{new_if};
+        mpl::log(mpl::Level::trace, category, "Adding new interface to cloud-init");
```

Comment:
The name of the function `add_extra_interfaces_to_cloud_init` is not clear enough. It is changed to `apply_extra_interfaces_to_cloud_init` in the PR above this one. 

It basically will overwrite the whole network-config file in cloud-init, which also means the input should the whole vector of existing extra interfaces. So the right call should be `instance->add_extra_interfaces_to_cloud_init(spec.default_mac_address, spec.extra_interfaces);`. Having a function which adds one single interface does make sense. However, when I was writing the function, I took a shortcut of reusing the function `make_cloud_init_network_config`, which eventually leads to the function `add_extra_interfaces_to_cloud_init` takes the whole extra interfaces vector. Anyway, for now we can apply the quick fix and keep it like this and see later on if we should change to single network interface addition.

Reply:
Yeah, this sounds good to me. Thanks!

Reply:
Changed.